[ota] Add an OTA library for SAM

henrikssn opened the pull request:
This allows the program to update itself, which is useful for OTA (on the air) updates of wireless sensors.

I tested it by flashing a blink sketch to address 0x10000, ran the example app and verified that LED started to blink and that data at *0x2000 and *0x12000 was equal to *0x10000.

henrikssn committed:
[ota] Add an OTA library for SAM

rleh commented:
Nice! :+1: 

Isn't this somehow the same feature as the "Flash driver" added by @salkinium in #417? Wouldn't it be nice to make the interface for STM32 and SAM the same here to be cross-platform portable?

salkinium commented:
I've recently implemented an almost identical updater. You can see [the application code here](https://gist.github.com/salkinium/fb0e4aef572d909c9ded1594dd081f1d), which uses the `modm:platform:flash` API for STM32 via #417. This code is identical for STM32G071 and STM32F446, even though the flash implementations are wildly different. Note that this uses an RPC via AMNB over UART, the call order is:

1. `start()` as a request/reponse to multiple individual nodes.
2. `fragment()` as a broadcast to multiple nodes.
3. `check()` as a request/reponse to multiple individual nodes to check for image completeness.
4. `apply()` as a broadcast to multiple nodes to update all at the same time (for example for baudrate changes).

The Flash API solves multiple issues:

- Complex alignment restrictions are left to the user to deal with. There's no good default way of doing it and the G0 family only does 64-bit updates, not 8-bit. You also cannot program the same flash address twice without erasure (so you cannot set 0 bits individually by programming the same address multiple times).
- Since the bus stalls on Flash erasure and program, you must relocate `Flash::erase` and `Flash::program` into `modm_ramcode` anyways.
- The `ram_apply` function is part of the user-code, since the `modm:platform:flash` module should also be useable for just storing non-volatile data, via some emulation layer and I wanted to keep the RAM code section small.
- The `ram_apply` function in user-code only calls `Flash::erase` and `Flash::program`, any other computations must be passed as arguments. This prevents the use of `modm_always_inline`, which is expensive code-size wise.
- The `modm:platform:flash` API is stateless, the elva::iap keeps track of what fragments/addresses have already been programmed.

I didn't add the `elva::iap` implementation to modm, since it's fairly specific to AMNB. But perhaps it makes sense to add it as  `modm:communication:amnb:iap` or something.
Similiarly for you the OTA implementation is obviously different, but a `modm:platform:flash` module for SAM should be usable by the ELVA IAP implementation too.

So do you want to split this implementation into a `modm:platform:flash` for SAM and some kind of IAP module (or leave it as example code for now? Note that there exists no platform independent interface yet, since I've only looked at STM32G0 and STM32F4 families, but there are more Flash implementations. (Also consider adding [an example](https://github.com/modm-io/modm/blob/develop/examples/nucleo_f446re/flash/main.cpp)).

PS: I called this an In-Application Programmer (IAP), not Bootloader or OTA, since you can communicate using different mediums too, and the upload happens during normal Application execution (unlike a Bootloader).

henrikssn commented:
Great minds think alike?

I'll give it a shot to break the library in two pieces, one which follows the existing flash platform pattern.

henrikssn committed:
[sam-iap] Migrate to use modm::platform::Flash

henrikssn commented:
So I did my best to shoehorn the SAM flash implementation into the existing flash driver. One complication I found is that SAM has a page buffer that needs to be written out in one step, but the existing Flash::program seems to write out a maximum of 32 bytes. I solved that by adding a `Flash::write()` method which is to be called by the user when one page has been written, and made it a no-op on STM32.

I moved the example into the generic folder and minimally modified it to handle the case mentioned above. Here are the speeds on my ATSAMD21E17 (Cortex-M0+) CPU.

```Erasing sectors [240, 480)
Erasing done in 241040us with errors: 0
Erasing with 261kiB/s
Programming done in 270730us with errors: 0
Programming with 232kiB/s
```

@salkinium would you mind patching this PR and testing the example on your devices?

Btw, I like @rleh's suggestion to make this a `modm::BlockDevice`, at least on SAM it should be quite easy to make that work, as the flash has a continuous address space. It would be useful to have the option to save the firmware updates on an external flash, thus being able to use the whole internal flash for firmware. I might go ahead and extend the SAM driver to (also) be a `BlockDevice`, but I'll leave the STM32 driver as an exercise to somebody else (that has access to hardware to test it on).

After that I only need to adapt the IAP library to match the new flash driver... I will take some inspiration from @salkinium's Gist here :)

salkinium commented on src/modm/platform/flash/sam/flash.hpp.in:
```diff
@@ -0,0 +1,78 @@
+	static uint32_t
+	erase(SectorType sector);
+
+	static inline uint32_t
```
This would probably need to be `modm_always_inline`, just to be sure.

salkinium commented on src/modm/platform/flash/sam/module.lb:
```diff
@@ -0,0 +1,39 @@
+
+    env.substitutions = {
+        "start": int(flash["start"], 16) + 0x2000,  # Adjust for bootloader.
+        "size": int(flash["size"]) - 0x2000,
```
Nah, don't check this here, do this in the IAP module.

salkinium commented:
> I solved that by adding a Flash::write() method which is to be called by the user when one page has been written, and made it a no-op on STM32.

I briefly considered adding a `Flash::programBlock(addr, count, type)` but you may not always want to write immediately, perhaps aggregate some fragments into the page buffer first, then separately `Flash::write` it. So I think this is a good solution.

> would you mind patching this PR and testing the example on your devices?

Works fine on STM32, you didn't change any functionality.

> It would be useful to have the option to save the firmware updates on an external flash, thus being able to use the whole internal flash for firmware.

It will be difficult moving all the code for that into RAM though. You will need a separate firmware compiled with a custom linkerscript to run from RAM alone, to copy the data from external to internal Flash. But that's basically a bootloader then.

> I might go ahead and extend the SAM driver to (also) be a BlockDevice

Tell me how it goes!

henrikssn committed:
[iap] Creates FirmwareImage class backed by BlockDevice

henrikssn commented:
Status after last commit:

- [x] Create images on hosted targets (/any targets supporting a BlockDevice)
- [x] Serialize/deserialize images from/to the underlying BlockDevice
- [x] Serialize/deserialize images in pieces (i.e. fixed size network packets)
- [x] Hosted unit tests for everything above
- [ ] Add a function to the Flash class that consumes a FirmwareImage and writes it to flash (overwriting the existing program code)

Was all of this strictly necessary to solve my original problem? Hell no but it was fun to write! :)

If anyone is doing something even remotely complex, I highly recommend making the code compile on hosted targets and setting up a unit test for it. The speedup in velocity is huge compared to debugging on the embedded device.

> > I solved that by adding a Flash::write() method which is to be called by the user when one page has been written, and made it a no-op on STM32.
> 
> I briefly considered adding a `Flash::programBlock(addr, count, type)` but you may not always want to write immediately, perhaps aggregate some fragments into the page buffer first, then separately `Flash::write` it. So I think this is a good solution.

Yup, I started using BlockDevice and eventually ran into this. I fixed it by creating buffered reader/writer classes for BlockDevice which takes care of this problem at the cost of some ram (64b on SAMD21).

If someone really feels that the RAM usage is a huge problem then you can template specialize the classes to utilize the ability of some flash to write partially to the page buffer. The advantage of the current solution is that it works with any block device, which is useful for testing on hosted targets and to actually build/send over the firmware image (i.e. compile on RPi and send it over UART).

> 
> > would you mind patching this PR and testing the example on your devices?
> 
> Works fine on STM32, you didn't change any functionality.
> 
> > It would be useful to have the option to save the firmware updates on an external flash, thus being able to use the whole internal flash for firmware.
> 
> It will be difficult moving all the code for that into RAM though. You will need a separate firmware compiled with a custom linkerscript to run from RAM alone, to copy the data from external to internal Flash. But that's basically a bootloader then.

You only need the Flash+BlockDevice classes running from RAM, I do think that is doable. I might need an lbuild param though to enable the ramcode functions only when IAP is needed, I'll see when I get there.

> 
> > I might go ahead and extend the SAM driver to (also) be a BlockDevice
> 
> Tell me how it goes!

The last commit is the current working draft, feel free to take a look if you are interested. Don't worry about all crap (e.g. commented out code) in there, I'll clean that up in a later commit.
